server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java

```diff
@@ -66,9 +65,13 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
 
     private final OpType opType;
 
+    @Nullable
+    private final DiscoveryNodeFilters trimmed;
+
     DiscoveryNodeFilters(OpType opType, Map<String, String[]> filters) {
         this.opType = opType;
-        this.filters = filters;
+        this.filters = Map.copyOf(filters);
+        this.trimmed = doTrim(this);
```

Comment:
Nit: could we avoid even calling this when constructing the trimmed filters? Seems unnecessary to walk the filters twice just to check that trimming actually worked.

Reply:
Actually  ... we can just avoid that walk altogether in `8` :) Missed that fact earlier, but we don't really need to walk the map in `8` because we're only filtering out a specific key (and `null` which I moved elsewhere but I think that filtering is unused ... just didn't want to change behavior here).
Simplified accordingly now